In [1]:
import pandas as pd
from pandas import read_csv, DataFrame

In [2]:
def count_cid_smiles(file_path):
    count = 0

    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file:
            count += 1
    print(f"Total entries (CID - SMILES): {count}")

file_path = "./Sources/CID-SMILES"  

count_cid_smiles(file_path)

Total entries (CID - SMILES): 122209523


`Remove Charged Species`

In [3]:
def remove_charged_species_smiles(input_path, output_path):
    kept = 0
    removed = 0

    with open(input_path, 'r', encoding='utf-8') as file, open(output_path, 'w', encoding='utf-8') as output:
        for line in file:
            parts = line.strip().split(None, 1)
            if len(parts) != 2:
                continue
            _, smiles = parts

            if '+' in smiles or '-' in smiles:
                removed += 1
                continue  

            output.write(line)
            kept += 1

    print(f"Removed SMILES (charged species): {removed}")
    print(f"Total SMILES: {kept}")

input_file = "./Sources/CID-SMILES"       
output_file = "./Sources/PubChemDatasets/CID-SMILES-NoChargedSpecies.txt"  

remove_charged_species_smiles(input_file, output_file)

Removed SMILES (charged species): 10049837
Total SMILES: 112159686


`Remove Metallic Compounds`

In [4]:
def remove_metallic_smiles(input_path, output_path):
    metals = [
        "Li", "Na", "K", "Rb", "Cs", "Fr",
        "Be", "Mg", "Ca", "Sr", "Ba", "Ra",
        "Fe", "Cu", "Zn", "Ni", "Au", "Ag", "Hg",
        "La", "Lu", "Ac", "Lr",
        "Al", "Ga", "In", "Sn", "Tl", "Pb", "Bi", "Po"
        "Sc", "Y", "Ti", "Zr", "Hf", "Rf",
        "V", "Nb", "Ta", "Db",
        "Cr", "Mo", "W", "Sg",
        "Mn", "Tc", "Re", "Bh",
        "Ru", "Os", "Hs",
        "Co", "Rh", "Ir", "Mt",
        "Pd", "Pt", "Ds",
        "Rg", "Cd", "Cn",
        "Nh", "Fl", "Mc", "Lv"
    ]
    
    kept = 0
    removed = 0

    with open(input_path, 'r', encoding='utf-8') as file, open(output_path, 'w', encoding='utf-8') as output:
        for line in file:
            parts = line.strip().split(None, 1)
            if len(parts) != 2:
                continue
            _, smiles = parts

            if any(metal in smiles for metal in metals):
                removed += 1
                continue

            output.write(line)
            kept += 1

    print(f"Removed SMILES (with metals): {removed}")
    print(f"Total SMILES: {kept}")

input_file = "./Sources/PubChemDatasets/CID-SMILES-NoChargedSpecies.txt"       
output_file = "./Sources/PubChemDatasets/CID-SMILES-NoMetallicCompounds.txt"             

remove_metallic_smiles(input_file, output_file)

Removed SMILES (with metals): 453961
Total SMILES: 111705725


`Remove Halogenated Compounds`

In [5]:
def remove_halogens_smiles(input_path, output_path):
    halogens = ["F", "Cl", "Br", "I", "At", "Ts"]
    kept = 0
    removed = 0

    with open(input_path, 'r', encoding='utf-8') as file, open(output_path, 'w', encoding='utf-8') as output:
        for line in file:
            parts = line.strip().split(None, 1)
            if len(parts) != 2:
                continue
            _, smiles = parts

            if any(halogen in smiles for halogen in halogens):
                removed += 1
                continue

            output.write(line)
            kept += 1

    print(f"Removed SMILES (halogens): {removed}")
    print(f"Total SMILES: {kept}")

input_file = "./Sources/PubChemDatasets/CID-SMILES-NoMetallicCompounds.txt"      
output_file = "./Sources/PubChemDatasets/CID-SMILES-NoHalogenatedCompounds.txt"  

remove_halogens_smiles(input_file, output_file)

Removed SMILES (halogens): 44816179
Total SMILES: 66889546


`Remove Mixtures SMILES`

In [6]:
def remove_mixtures_smiles(input_path, output_path):
    kept = 0
    removed = 0

    with open(input_path, 'r', encoding='utf-8') as file, open(output_path, 'w', encoding='utf-8') as output:
        for line in file:
            parts = line.strip().split(None, 1)
            if len(parts) != 2:
                continue
            _, smiles = parts

            if '.' in smiles:
                removed += 1
                continue  

            output.write(line)
            kept += 1

    print(f"Removed SMILES (with mixtures): {removed}")
    print(f"Total SMILES: {kept}")

input_file = "./Sources/PubChemDatasets/CID-SMILES-NoHalogenatedCompounds.txt"  
output_file = "./Sources/PubChemDatasets/CID-SMILES-NoMixtures.txt"  

remove_mixtures_smiles(input_file, output_file)

Removed SMILES (with mixtures): 3246005
Total SMILES: 63643541


`Attain Compounds with Melting and Boiling Temperature`

In [7]:
from chemicals.phase_change import Tm_sources

for source, df in Tm_sources.items ():
    print (f"\nSource: {source}")
    print (f"Columns: {df.columns.tolist()}")


Source: OPEN_NTBKM
Columns: ['Tm']

Source: CRC_INORG
Columns: ['Chemical', 'Tm', 'Tb', 'rho']

Source: CRC_ORG
Columns: ['Name', 'Tm', 'Tb', 'rho', 'RI']

Source: COMMON_CHEMISTRY
Columns: ['Tm', 'Tb', 'Vms', 'Vml']

Source: WEBBOOK
Columns: ['Tt', 'Tm', 'Tb', 'Tc', 'Pt', 'Pc', 'Vc', 'Hfs', 'Hfl', 'Hfg', 'S0s', 'S0l', 'S0g', 'Hsub', 'Hfus', 'Hvap', 'AntoineTmin', 'AntoineTmax', 'AntoineA', 'AntoineB', 'AntoineC']

Source: WIKIDATA
Columns: ['Tm', 'Tb', 'Hfus', 'Hvap', 'Hf', 'S0', 'LFL', 'UFL', 'T_flash', 'T_autoignition', 'RI', 'RIT', 'logP']

Source: JOBACK
Columns: ['Tm', 'Hfus', 'Hvap', 'Tb', 'Tc', 'Pc', 'Vc', 'Hfg', 'Cpg0', 'Cpg1', 'Cpg2', 'Cpg3', 'mul0', 'mul1']


In [8]:
from chemicals.phase_change import Tb_sources

for source, df in Tb_sources.items ():
    print (f"\nSource: {source}")
    print (f"Columns: {df.columns.tolist()}")


Source: HEOS
Columns: ['ID', 'name', 'Tc', 'Pc', 'Vc', 'Tt', 'Pt', 'Tb', 'omega']

Source: CRC_ORG
Columns: ['Name', 'Tm', 'Tb', 'rho', 'RI']

Source: CRC_INORG
Columns: ['Chemical', 'Tm', 'Tb', 'rho']

Source: COMMON_CHEMISTRY
Columns: ['Tm', 'Tb', 'Vms', 'Vml']

Source: WEBBOOK
Columns: ['Tt', 'Tm', 'Tb', 'Tc', 'Pt', 'Pc', 'Vc', 'Hfs', 'Hfl', 'Hfg', 'S0s', 'S0l', 'S0g', 'Hsub', 'Hfus', 'Hvap', 'AntoineTmin', 'AntoineTmax', 'AntoineA', 'AntoineB', 'AntoineC']

Source: YAWS
Columns: ['Tb']

Source: WIKIDATA
Columns: ['Tm', 'Tb', 'Hfus', 'Hvap', 'Hf', 'S0', 'LFL', 'UFL', 'T_flash', 'T_autoignition', 'RI', 'RIT', 'logP']

Source: JOBACK
Columns: ['Tm', 'Hfus', 'Hvap', 'Tb', 'Tc', 'Pc', 'Vc', 'Hfg', 'Cpg0', 'Cpg1', 'Cpg2', 'Cpg3', 'mul0', 'mul1']


### Data Sources

- **`CRC_ORG`**  
  Compilation of evaluated data on organic compounds from the CRC Handbook of Chemistry and Physics [1].

- **`CRC_INORG`**  
  Compilation of evaluated data on inorganic compounds from the CRC Handbook of Chemistry and Physics [1].

- **`YAWS`**  
  Compilation of experimental and predicted data aggregated from multiple sources, without individual sources attribution [2].

- **`WIKIDATA`**  
  Property values retrieved from the collaborative Wikidata knowledge base via its public API [3].

- **`COMMON_CHEMISTRY`**  
  Curated data from the CAS Common Chemistry project [4].

- **`JOBACK`**  
  Properties estimated for organic compounds via the Joback-Reid group-contribution method [5].

- **`WEBBOOK`**  
  Compiled experimental thermochemical and thermophysical data from the NIST Chemistry WebBook [6].

- **`HEOS`**  
  Thermophysical properties computed from fundamental Helmholtz-energy equations of state, as distributed in the NIST REFPROP Database [7].

### References

[1] Haynes, W.M., Thomas J. Bruno, and David R. Lide. CRC Handbook of Chemistry and Physics, 95E. Boca Raton, FL: CRC Press, 2014.

[2] Yaws, Carl L. Thermophysical Properties of Chemicals and Hydrocarbons, Second Edition. Amsterdam Boston: Gulf Professional Publishing, 2014.

[3] Wikidata. Accessed via API: [https://www.wikidata.org/](https://www.wikidata.org/)

[4] CAS Common Chemistry: [https://commonchemistry.cas.org/](https://commonchemistry.cas.org/)

[5] Joback, K.G., and R.C. Reid. "Estimation of Pure-Component Properties from Group-Contributions." Chemical Engineering Communications 57, no. 1-6 (1987): 233–243. [https://doi.org/10.1080/00986448708960487](https://doi.org/10.1080/00986448708960487)

[6] Shen, V.K., Siderius, D.W., Krekelberg, W.P., and Hatch, H.W., Eds. NIST WebBook, NIST. [https://doi.org/10.18434/T4M88Q](https://doi.org/10.18434/T4M88Q)

[7] Huber, Marcia L., Eric W. Lemmon, Ian H. Bell, and Mark O. McLinden. "The NIST REFPROP Database for Highly Accurate Properties of Industrially Important Fluids." Industrial & Engineering Chemistry Research 61, no. 42 (2022): 15449–15472. [https://doi.org/10.1021/acs.iecr.2c01427](https://doi.org/10.1021/acs.iecr.2c01427)


- Retrieve all melting points from sources

In [9]:
mp_columns = ['Tm']
records = []

for source, df in Tm_sources.items():
    mp_col = next((col for col in mp_columns if col in df.columns), None)
   
    if 'CASRN' in df.columns:
        result_df = df[['CASRN', mp_col]].copy()
    else:
        result_df = df[[mp_col]].copy()
        result_df.reset_index(inplace=True)
        result_df.rename(columns={'index': 'CASRN'}, inplace=True)

    result_df.columns = ['CASRN', 'Tm']
    result_df['Source'] = source
    records.append(result_df)

all_melting_points = pd.concat(records, ignore_index=True)
all_melting_points = all_melting_points.dropna(subset=['Tm'])
all_melting_points = all_melting_points.drop_duplicates(subset='CASRN')
print(f"Total unique melting point records: {len(all_melting_points)}")

Total unique melting point records: 111389


- Retrieve all boiling points from sources

In [10]:
bp_columns = ['Tb', 'T', 'T_b']
records = []

for source, df in Tb_sources.items():
    bp_col = next((col for col in bp_columns if col in df.columns), None)
    
    if 'CASRN' in df.columns:
        result_df = df[['CASRN', bp_col]].copy()
    else:
        result_df = df[[bp_col]].copy()
        result_df.reset_index(inplace=True)
        result_df.rename(columns={'index': 'CASRN'}, inplace=True)

    result_df.columns = ['CASRN', 'Tb'] 
    result_df['Source'] = source
    records.append(result_df)

all_boiling_points = pd.concat(records, ignore_index=True)
all_boiling_points = all_boiling_points.dropna(subset=['Tb'])
all_boiling_points = all_boiling_points.drop_duplicates(subset='CASRN')
print(f"Total unique boiling point records: {len(all_boiling_points)}")

Total unique boiling point records: 70227


In [ ]:
#all_melting_points.to_csv("./all_melting_points_source.csv", index = False)
#all_boiling_points.to_csv("./all_boiling_points_source.csv", index = False)

In [12]:
filename = "./Sources/Temperatures/all_melting_points_source.csv"
all_melting_points: DataFrame = read_csv (filename, na_values="")
all_melting_points.shape

(111389, 3)

In [14]:
filename = "./Sources/Temperatures/all_boiling_points_source.csv"
all_boililng_points: DataFrame = read_csv (filename, na_values="")
all_boiling_points.shape

(70227, 3)

In [15]:
all_boiling_melting_combined = pd.merge(
    all_boililng_points[['CASRN', 'Tb']],  
    all_melting_points[['CASRN', 'Tm']], 
    on='CASRN',
    how='inner'  
)
all_boiling_melting_combined.shape

(56787, 3)

In [16]:
def format_cas_number(raw_cas):
    digits = ''.join(filter(str.isdigit, str(raw_cas)))
    
    if len(digits) < 3:
        return None 
    
    part3 = digits[-1]
    part2 = digits[-3:-1]
    part1 = digits[:-3]
    
    return f"{part1}-{part2}-{part3}"

In [17]:
all_boiling_melting_combined["CAS"] = all_boiling_melting_combined["CASRN"].apply(format_cas_number)
print(all_boiling_melting_combined['CAS'].nunique()) 

51842


In [18]:
#all_boiling_melting_combined.to_csv("./Sources/Temperatures/all_boiling_melting_combined.csv", index = False)

`Attain PubChem CID`

In [19]:
import pubchempy as pcp
import re

In [47]:
def get_cids_from_cas(cas_list):
    cid_results = {}
    total = len(cas_list)
    for i, cas in enumerate(cas_list, 1):
        try:
            cids = pcp.get_cids(cas, 'name')
            if cids:
                cid_results[cas] = cids[0]
            else:
                cid_results[cas] = None
        except Exception:
            cid_results[cas] = None

        cid = cid_results[cas]
        cid_str = f"CID: {cid}" if cid else "CID not found"
        print(f"[{i}/{total}] CAS: {cas} - {cid_str}")

    return cid_results

In [ ]:
cas_list = all_boiling_melting_combined['CAS'].dropna().tolist()
#cid_results = get_cids_from_cas(cas_list)

In [22]:
filename = "./Sources/Temperatures/all_boiling_melting_combined.csv"

pattern = re.compile(r"CAS:\s*(\S+?)\s+\S+\s+(?:CID:\s*(\d+)|CID not found)")
cas_to_cid = {}
with open("./Sources/PubChemDatasets/CAS-CID.txt", "r", encoding="utf-8") as f:
    for line in f:
        m = pattern.search(line)
        if m:
            cas_to_cid[m.group(1)] = int(m.group(2)) if m.group(2) else None

dataset = pd.read_csv(filename, na_values= "")
dataset["CID"]    = dataset["CAS"].map(cas_to_cid).astype("Int64")
dataset["SMILES"] = ""

dataset_results = dataset[["CAS", "CID", "SMILES", "Tb", "Tm"]]
#dataset_results.to_csv("./Sources/SolventsDatasets/dataset_CAS_CID_Tb_Tm.csv", index=False)

`Attain SMILES from PubChem`

In [23]:
filename = "./Sources/SolventsDatasets/dataset_CAS_CID_Tb_Tm.csv"
dataset_results: DataFrame = read_csv (filename, na_values="")
dataset_results.shape

(56787, 5)

- Remove compounds without CID

In [24]:
dataset_results_smiles = dataset_results.dropna(subset = ['CID'])
dataset_results_smiles.shape

(55091, 5)

In [ ]:
cids_to_find = set(dataset_results_smiles['CID'].dropna())

cid_to_smiles = {}
total_lines = 0
found_smiles = 0

with open('./Sources/PubChemDatasets/CID-SMILES-NoMixtures.txt', 'r') as f:
    for line in f:
        total_lines += 1
        line = line.strip()
        if not line:
            continue

        parts = line.split(maxsplit=1)
        if len(parts) != 2 or not parts[0].isdigit():
            continue

        cid = int(parts[0])
        smiles = parts[1]

        if cid in cids_to_find:
            cid_to_smiles[cid] = smiles
            found_smiles += 1

dataset_results_smiles['SMILES'] = dataset_results_smiles['CID'].map(cid_to_smiles)

print(f"Total SMILES found: {found_smiles:,}")
print(f"Total CAS - CID entries: {len(dataset_results_smiles):,}")
print(f"Missing SMILES: {dataset_results_smiles['SMILES'].isna().sum():,}")

dataset_results_smiles['CID'] = dataset_results_smiles['CID'].astype("Int64")
#dataset_results_smiles.to_csv("./Sources/SolventsDatasets/dataset_CAS_CID_SMILES_Tb_Tm.csv", index = False)

Total SMILES found: 33318

Total CAS - CID entries: 55091

Missing SMILES: 18226

`Filter by SMILES`

In [26]:
filename = "./Sources/SolventsDatasets/dataset_CAS_CID_SMILES_Tb_Tm.csv"
data: DataFrame = read_csv (filename, delimiter = ",")
data.shape

(55091, 5)

- Update Water SMILES

In [27]:
search_value = '7732-18-5'
result = data[data['CAS'] == search_value]

if not result.empty:
    smiles_value = result.iloc[0]['SMILES'] 
    data.loc[data['CAS'] == search_value, 'SMILES'] = 'O'
    print("SMILES value updated to 'O'")

SMILES value updated to 'O'


- Remove Missing SMILES

In [28]:
missing_smiles = data[data['SMILES'].isna()]['CAS']
len(missing_smiles)

18226

In [29]:
data = data.dropna(subset=['SMILES'])
data.shape

(36865, 5)

`Keep compounds with H, C, N, O, S`

In [30]:
allowed_elements = {'H', 'C', 'N', 'O', 'S'}

def contains_only_HCNOS(smiles):
    elements = re.findall(r'\[.*?\]|Br|Cl|[A-Z][a-z]?', smiles)

    cleaned_elements = []
    for el in elements:
        if el.startswith('['):
            el = el.strip('[]')
            sub_elements = re.findall(r'[A-Z][a-z]?', el)
            cleaned_elements.extend(sub_elements)
        else:
            cleaned_elements.append(el)

    return all(el in allowed_elements for el in cleaned_elements)

data = data[data['SMILES'].apply(contains_only_HCNOS)].copy()
data.shape

(36415, 5)

`Filter by Temperature` - Keep compounds with Tm bellow or equal to 333.15 K and Tb above or equal to 333.15 K 

In [31]:
data  = data[(data['Tm'] <= 333.15) & (data['Tb'] >= 333.15)].copy()
data.shape

(16069, 5)

- Duplicated instances occuring from multiple Tm and Tb sources 

In [32]:
data_duplicates = data[data['CAS'].duplicated(keep=False)].copy()
print(f"Number of records: {len(data)}")
original_number_records = len(data)
print(f"Number of duplicated records: {len(data_duplicates)}")

Number of records: 16069
Number of duplicated records: 4322


In [33]:
def format_cas_number(raw_cas):
    digits = ''.join(filter(str.isdigit, str(raw_cas)))
    
    if len(digits) < 3:
        return None 
    
    part3 = digits[-1]
    part2 = digits[-3:-1]
    part1 = digits[:-3]
    
    return f"{part1}-{part2}-{part3}"

In [34]:
filename = "./Sources/Temperatures/all_melting_points_source.csv"
all_melting_points: DataFrame = read_csv (filename, na_values="")
all_melting_points.rename(columns={'CASRN': 'CAS'}, inplace=True)
all_melting_points.shape

(111389, 3)

In [35]:
filename = "./Sources/Temperatures/all_boiling_points_source.csv"
all_boiling_points: DataFrame = read_csv (filename, na_values="")
all_boiling_points.rename(columns={'CASRN': 'CAS'}, inplace=True)
all_boiling_points.shape

(70227, 3)

In [36]:
all_boiling_points['CAS'] = all_boiling_points['CAS'].apply(format_cas_number)
all_melting_points['CAS'] = all_melting_points['CAS'].apply(format_cas_number)

- Label sources of the individual melting and boiling temperature values

In [37]:
tb_lookup = dict(zip(all_boiling_points['Tb'], all_boiling_points['Source']))
tm_lookup = dict(zip(all_melting_points['Tm'], all_melting_points['Source']))

def match_tb(row):
    return tb_lookup.get(row['Tb'], None)

def match_tm(row):
    return tm_lookup.get(row['Tm'], None)

data['Tb Source'] = data.apply(match_tb, axis=1)
data['Tm Source'] = data.apply(match_tm, axis=1)

`1) Filter by Temperature` - Keep records from CRC_ORG or CRC_INORG

In [38]:
EXPERIMENTAL_PRIORITY = {'CRC_ORG', 'CRC_INORG'}

def filter_prefer_crc(df):
    removed = 0
    kept_rows = []

    for _, group in df.groupby('CAS'):
        if len(group) == 1:
            kept_rows.append(group)
            continue

        crc_rows = group[
            group['Tb Source'].isin(EXPERIMENTAL_PRIORITY) |
            group['Tm Source'].isin(EXPERIMENTAL_PRIORITY)
        ]
        if not crc_rows.empty:
            removed += len(group) - 1
            kept_rows.append(crc_rows.iloc[[0]])
        else:
            kept_rows.append(group)

    return pd.concat(kept_rows), removed

data, removed_crc = filter_prefer_crc(data)
print(f"Current Number records: {len(data)} vs Original Number records: {original_number_records}")
print(f"Removed records: {removed_crc}")

Current Number records: 15943 vs Original Number records: 16069
Removed records: 126


` 2) Filter by Temperature` - If only computed data exists (JOBACK) keep first instance

In [39]:
def filter_all_joback_only(df):
    removed = 0
    kept_rows = []

    for _, group in df.groupby('CAS'):
        if (group['Tb Source'] == 'JOBACK').all() and (group['Tm Source'] == 'JOBACK').all():
            removed += len(group) - 1
            kept_rows.append(group.iloc[[0]])
        else:
            kept_rows.append(group)
    
    return pd.concat(kept_rows), removed

data, removed_joback = filter_all_joback_only(data)
print(f"Current Number records: {len(data)} vs Original Number records: {original_number_records}")
print(f"Removed records: {removed_joback}")

Current Number records: 15767 vs Original Number records: 16069
Removed records: 176


`3) Filter by Temperature` - Keep only first instance

In [40]:
def filter_keep_first_only(df):
    removed = 0
    kept_rows = []

    for _, group in df.groupby('CAS'):
        if len(group) > 1:
            removed += len(group) - 1
        kept_rows.append(group.iloc[[0]])

    return pd.concat(kept_rows), removed

data, removed_final = filter_keep_first_only(data)
print(f"Current Number records: {len(data)} vs Original Number records: {original_number_records}")
print(f"Removed records: {removed_final}")

Current Number records: 13908 vs Original Number records: 16069
Removed records: 1859


- Verify all duplicated entries were removed

In [41]:
num_remaining_duplicates = data['CAS'].duplicated().sum()
print(f"Remaining duplicated CAS entries: {num_remaining_duplicates}")
print(f"Number of total records removed: {removed_joback + removed_final + removed_crc} ")
print(f"Current Number records: {len(data)}")

Remaining duplicated CAS entries: 0
Number of total records removed: 2161 
Current Number records: 13908


`Filter by Functional Groups` - RDKit Package

In [42]:
from rdkit import Chem
from rdkit.Chem import Fragments

In [43]:
EXCLUDED_FRAGMENTS = [
    Fragments.fr_C_S,           
    Fragments.fr_Imine,         
    Fragments.fr_NH1,           
    Fragments.fr_NH2,           
    Fragments.fr_SH,            
    Fragments.fr_aniline,       
    Fragments.fr_azide,         
    Fragments.fr_azo,           
    Fragments.fr_barbitur,      
    Fragments.fr_benzodiazepine,
    Fragments.fr_diazo,
    Fragments.fr_epoxide,
    Fragments.fr_furan,
    Fragments.fr_guanido,
    Fragments.fr_hdrzine,       
    Fragments.fr_hdrzone,       
    Fragments.fr_imidazole,
    Fragments.fr_imide,
    Fragments.fr_isocyan,       
    Fragments.fr_isothiocyan,
    Fragments.fr_nitro,
    Fragments.fr_nitro_arom,
    Fragments.fr_nitro_arom_nonortho,
    Fragments.fr_nitroso,
    Fragments.fr_oxazole,
    Fragments.fr_oxime,
    Fragments.fr_phos_acid,
    Fragments.fr_phos_ester,
    Fragments.fr_piperzine,
    Fragments.fr_piperdine,
    Fragments.fr_prisulfonamd,
    Fragments.fr_sulfide,
    Fragments.fr_sulfonamd,
    Fragments.fr_term_acetylene,
    Fragments.fr_tetrazole,
    Fragments.fr_thiazole,
    Fragments.fr_thiocyan,
    Fragments.fr_thiophene,
    Fragments.fr_urea,
    Fragments.fr_N_O,
    Fragments.fr_Nhpyrrole,
    Fragments.fr_alkyl_carbamate,
    Fragments.fr_lactam,
    Fragments.fr_morpholine,
    Fragments.fr_quatN,
    Fragments.fr_aldehyde,
    Fragments.fr_amidine,
    Fragments.fr_Ar_NH,
    Fragments.fr_dihydropyridine     
]

def has_excluded_fragments(smiles):
    mol = Chem.MolFromSmiles(smiles)
    return any(fragment(mol) > 0 for fragment in EXCLUDED_FRAGMENTS)

initial = len(data)
df_filtered = data[~data['SMILES'].apply(has_excluded_fragments)].copy()
print(f"Before: {initial} | After: {len(df_filtered)} | Removed: {initial - len(df_filtered)}")

Before: 13908 | After: 10382 | Removed: 3526


`Filter by Adjacent Functional Groups` - SMARTS

In [44]:
excluded_patterns = {"adjacent O-O" : Chem.MolFromSmarts("[O][O]"), "adjacent S-S" : Chem.MolFromSmarts("[S][S]")}

def has_excluded_patterns(smiles):
    mol = Chem.MolFromSmiles(smiles)
    return any(mol.HasSubstructMatch(pattern)for pattern in excluded_patterns.values())

initial = len(df_filtered)
df_results = df_filtered[~df_filtered['SMILES'].apply(has_excluded_patterns)].copy()
print(f"Before: {initial} | After: {len(df_results)} | Removed: {initial - len(df_results)}")               

Before: 10382 | After: 10204 | Removed: 178


In [45]:
#df_results.to_csv("./Sources/SolventsDatasets/dataset_CAS_CID_SMILES_Tm_Tb_Sources.csv", index = False)
df_results.shape

(10204, 7)